Convert node-wide request rate limiter to a per-channel outstanding request limiter.

clemahieu opened the pull request:
This patch adds channel scoring functionality to the ascending bootstrapper. It prioritises using the most responsive channels first, e.g. the channels with the lowest outstanding requests.

Since there is no guarantee requests will be responded to, the scores periodically "decay" when timeouts are calculated.

This scoring opens up future possibilities for additional metrics; an example would be tracking telemetry information and not requesting nodes that report low block counts.

Performance wise this looks to have ~2x improvement over randomly selecting channels as was done previously.

thsfs commented on nano/node/bootstrap_ascending/peer_scoring.cpp:
```diff
@@ -0,0 +1,127 @@
+	for (auto i = index.begin (), n = index.end (); i != n;)
+	{
+		auto & score = *i;
+		//std::cerr << boost::str (boost::format ("%1%: %2% %3% %4%\n") % score.endpoint % score.outstanding % score.request_count_total % score.response_count_total);
```
These "std::cerr" output comments can be removed.

thsfs commented on nano/node/bootstrap_ascending/peer_scoring.cpp:
```diff
@@ -0,0 +1,127 @@
+void nano::bootstrap_ascending::peer_scoring::timeout ()
+{
+	auto & index = scoring.get<tag_channel> ();
+	for (auto i = index.begin (), n = index.end (); i != n;)
```
suggest naming the "i" variable as "peer_score" for better readability.

thsfs commented on nano/node/bootstrap_ascending/peer_scoring.cpp:
```diff
@@ -0,0 +1,127 @@
+	auto & index = scoring.get<tag_channel> ();
+	for (auto i = index.begin (), n = index.end (); i != n;)
+	{
+		auto & score = *i;
```
Not used.

thsfs commented on nano/node/bootstrap_ascending/peer_scoring.cpp:
```diff
@@ -0,0 +1,127 @@
+void nano::bootstrap_ascending::peer_scoring::sync (std::deque<std::shared_ptr<nano::transport::channel>> const & list)
+{
+	auto & index = scoring.get<tag_channel> ();
+	for (auto channel : list)
```
channel can be a const reference.

thsfs commented on nano/node/bootstrap_ascending/peer_scoring.cpp:
```diff
@@ -109,9 +113,24 @@ void nano::bootstrap_ascending::peer_scoring::sync (std::deque<std::shared_ptr<n
 				if (!channel->max (nano::transport::traffic_type::bootstrap))
 				{
-					index.insert ({ channel, channel.get (), channel->get_endpoint (), 1, 1, 0 });
+					index.emplace (channel, 1, 1, 0);
```
The counters could go also as default values in the constructor, not sure which way is better.

thsfs commented on nano/node/bootstrap_ascending/peer_scoring.cpp:
```diff
@@ -0,0 +1,136 @@
+	return false;
+}
+
+void nano::bootstrap_ascending::peer_scoring::received_message (std::shared_ptr<nano::transport::channel> channel)
```
This function has no use yet.

clemahieu replied:
This was missed when rebasing.

thsfs commented on nano/node/bootstrap_ascending/peer_scoring.hpp:
```diff
@@ -0,0 +1,88 @@
+			// std::weak_ptr does not provide ordering so the naked pointer is also tracked and used for ordering channels
+			// This pointer may be invalid if the channel has been destroyed
+			nano::transport::channel * channel_ptr;
+			nano::endpoint endpoint;
```
The endpoint has no use yet and it can be retrieved from the channel.

clemahieu committed:
Convert node-wide request rate limiter to a per-channel outstanding request limiter.
Adding prioritizing for selecting channels, picking most responsive channel first.
